Add index and hash JMH benchmarks

DanielThomas opened the pull request:
I needed to make some decisions about how to best model/index data for a performance sensitive use case and rather than go through trial and error in the actual implementation I figured I'd add some benchmarks here instead.

The full suite has a lot of permutations, so would several hours to run, so I cherry picked a couple of configurations that gives a meaningful comparison. Would appreciate some feedback on the methodology to make sure the results are correct.

Our use case would currently have ~65 million objects, with cardinality ranging from 1 to 100 thousand - the main questions I was trying to answer here were:

- Nested or non-nested objects for index fields
- Compound or non-compound keys
- Should we seek to reduce cardinality by adding separate collection holder type and group them in the consumer, or is `HashIndex` creation/load performance good enough in the consumer

# HashIndex

`cardinality` refers to the number of matches (i.e. number of objects with identical fields).

```
Benchmark                                    (cardinality)  (nested)  (querySize)   (size)  Mode  Cnt    Score   Error  Units
HollowHashIndexBenchmark.createIndex                     1     false            1  1000000  avgt         0.470           s/op
HollowHashIndexBenchmark.createIndex                     1     false            4  1000000  avgt         0.511           s/op
HollowHashIndexBenchmark.createIndex                     1      true            1  1000000  avgt         0.472           s/op
HollowHashIndexBenchmark.createIndex                     1      true            4  1000000  avgt         0.555           s/op
HollowHashIndexBenchmark.findMatches                     1     false            1  1000000  avgt       369.577          ns/op
HollowHashIndexBenchmark.findMatches                     1     false            4  1000000  avgt       407.978          ns/op
HollowHashIndexBenchmark.findMatches                     1      true            1  1000000  avgt       620.034          ns/op
HollowHashIndexBenchmark.findMatches                     1      true            4  1000000  avgt       642.681          ns/op
HollowHashIndexBenchmark.findMatchesMissing              1     false            1  1000000  avgt        46.252          ns/op
HollowHashIndexBenchmark.findMatchesMissing              1     false            4  1000000  avgt        35.563          ns/op
HollowHashIndexBenchmark.findMatchesMissing              1      true            1  1000000  avgt        64.633          ns/op
HollowHashIndexBenchmark.findMatchesMissing              1      true            4  1000000  avgt        35.785          ns/op
```

Performance improves as cardinality increases:
```
Benchmark                                    (cardinality)  (nested)  (querySize)   (size)  Mode  Cnt    Score   Error  Units
HollowHashIndexBenchmark.createIndex                  1000     false            1  1000000  avgt         0.001           s/op
HollowHashIndexBenchmark.createIndex                  1000     false            4  1000000  avgt         0.002           s/op
HollowHashIndexBenchmark.createIndex                  1000      true            1  1000000  avgt         0.001           s/op
HollowHashIndexBenchmark.createIndex                  1000      true            4  1000000  avgt         0.002           s/op
HollowHashIndexBenchmark.findMatches                  1000     false            1  1000000  avgt        65.274          ns/op
HollowHashIndexBenchmark.findMatches                  1000     false            4  1000000  avgt       133.855          ns/op
HollowHashIndexBenchmark.findMatches                  1000      true            1  1000000  avgt        77.624          ns/op
HollowHashIndexBenchmark.findMatches                  1000      true            4  1000000  avgt       221.643          ns/op
HollowHashIndexBenchmark.findMatchesMissing           1000     false            1  1000000  avgt        22.621          ns/op
HollowHashIndexBenchmark.findMatchesMissing           1000     false            4  1000000  avgt        34.524          ns/op
HollowHashIndexBenchmark.findMatchesMissing           1000      true            1  1000000  avgt        22.642          ns/op
HollowHashIndexBenchmark.findMatchesMissing           1000      true            4  1000000  avgt        34.178          ns/op
```

# PrimaryKeyIndex

```
Benchmark                                                 (nested)  (querySize)   (size)  Mode  Cnt    Score   Error  Units
HollowPrimaryKeyIndexBenchmark.createIndex                   false            1  1000000  avgt         0.063           s/op
HollowPrimaryKeyIndexBenchmark.createIndex                   false            4  1000000  avgt         0.095           s/op
HollowPrimaryKeyIndexBenchmark.createIndex                    true            1  1000000  avgt         0.116           s/op
HollowPrimaryKeyIndexBenchmark.createIndex                    true            4  1000000  avgt         0.189           s/op
HollowPrimaryKeyIndexBenchmark.getMatchingOrdinal            false            1  1000000  avgt       305.982          ns/op
HollowPrimaryKeyIndexBenchmark.getMatchingOrdinal            false            4  1000000  avgt       333.696          ns/op
HollowPrimaryKeyIndexBenchmark.getMatchingOrdinal             true            1  1000000  avgt       563.586          ns/op
HollowPrimaryKeyIndexBenchmark.getMatchingOrdinal             true            4  1000000  avgt       595.999          ns/op
HollowPrimaryKeyIndexBenchmark.getMatchingOrdinalMissing     false            1  1000000  avgt        40.421          ns/op
HollowPrimaryKeyIndexBenchmark.getMatchingOrdinalMissing     false            4  1000000  avgt        36.977          ns/op
HollowPrimaryKeyIndexBenchmark.getMatchingOrdinalMissing      true            1  1000000  avgt        73.481          ns/op
HollowPrimaryKeyIndexBenchmark.getMatchingOrdinalMissing      true            4  1000000  avgt        41.947          ns/op
```

# HashCodes

I went with int keys in the benchmarks, to avoid the hash function being a significant contributor. Added some benchmarks to look at the hash functions separately. You might want to look at the `String` variant of `hashCode`, there's some weirdness about multibyte Strings and has worst performance than hashing a byte array of the same length:
```
Benchmark                                (length)  (multibyte)  Mode  Cnt     Score     Error  Units
Benchmark                               (length)  Mode  Cnt     Score   Error  Units
HashCodesBenchmark.hashBytes                   1  avgt          7.247          ns/op
HashCodesBenchmark.hashBytes                   2  avgt          7.857          ns/op
HashCodesBenchmark.hashBytes                   3  avgt          7.871          ns/op
HashCodesBenchmark.hashBytes                   5  avgt          8.933          ns/op
HashCodesBenchmark.hashBytes                  10  avgt         12.028          ns/op
HashCodesBenchmark.hashBytes                 100  avgt         56.531          ns/op
HashCodesBenchmark.hashBytes                1000  avgt        524.596          ns/op
HashCodesBenchmark.hashInt                     1  avgt          2.630          ns/op
HashCodesBenchmark.hashInt                     2  avgt          2.671          ns/op
HashCodesBenchmark.hashInt                     3  avgt          2.656          ns/op
HashCodesBenchmark.hashInt                     5  avgt          2.656          ns/op
HashCodesBenchmark.hashInt                    10  avgt          2.638          ns/op
HashCodesBenchmark.hashInt                   100  avgt          2.633          ns/op
HashCodesBenchmark.hashInt                  1000  avgt          2.628          ns/op
HashCodesBenchmark.hashLong                    1  avgt          2.705          ns/op
HashCodesBenchmark.hashLong                    2  avgt          2.701          ns/op
HashCodesBenchmark.hashLong                    3  avgt          2.715          ns/op
HashCodesBenchmark.hashLong                    5  avgt          2.731          ns/op
HashCodesBenchmark.hashLong                   10  avgt          2.737          ns/op
HashCodesBenchmark.hashLong                  100  avgt          2.715          ns/op
HashCodesBenchmark.hashLong                 1000  avgt          2.736          ns/op
HashCodesBenchmark.hashString                  1  avgt          9.322          ns/op
HashCodesBenchmark.hashString                  2  avgt         10.185          ns/op
HashCodesBenchmark.hashString                  3  avgt         10.487          ns/op
HashCodesBenchmark.hashString                  5  avgt         13.324          ns/op
HashCodesBenchmark.hashString                 10  avgt         18.794          ns/op
HashCodesBenchmark.hashString                100  avgt         98.877          ns/op
HashCodesBenchmark.hashString               1000  avgt        868.692          ns/op
HashCodesBenchmark.hashStringMultibyte         1  avgt         24.457          ns/op
HashCodesBenchmark.hashStringMultibyte         2  avgt         26.278          ns/op
HashCodesBenchmark.hashStringMultibyte         3  avgt         30.154          ns/op
HashCodesBenchmark.hashStringMultibyte         5  avgt         31.497          ns/op
HashCodesBenchmark.hashStringMultibyte        10  avgt         43.098          ns/op
HashCodesBenchmark.hashStringMultibyte       100  avgt        286.791          ns/op
HashCodesBenchmark.hashStringMultibyte      1000  avgt       2789.318          ns/op
```

DanielThomas committed:
Add index related benchmarks

DanielThomas committed:
Fix import order

ghost commented:
## Pre-compute

**Update:** removed my incorrect advice to use `Map` or `Set` as a pre-computed hash index. You can, however, use that technique to pre-compute a *primary key* index. See [map schemas](https://hollow.how/data-modeling/#map-schemas) and [set schemas](https://hollow.how/data-modeling/#set-schemas).

## HashIndex re-indexing

The current hash index implementation has a naive implementation when a delta is applied. Where primary/unique key indices take into account the added/deleted/modified records and only re-index those, hash index currently re-indexes everything on any update (delta or snapshot).

All indices do a full re-index on snapshot (either initial load or double-snapshot updates), so this only matters when your producer consistently sends deltas (which within Netflix is common).

PaulSandoz reviewed (approved):
Very nice, thank you for contributing. This is very timely as i was going to add some JMH benchmarks but had not worked out the Gradle integration.

PaulSandoz commented:
Discussing with Danny we think it's best to spin up a separate repo for Hollow JMH benchmarks. I'll take the contents from this pull request as a starting point and merge in my JMH benchmarks as the initial commit to this new repo.

rpalcolea commented:
This is super nice! definitely one of the wish-list 👍

DanielThomas committed:
Reflect actual performance of low cardinality indexes

I realised that small indexes are going to just sit in cache in a tight loop and the actual performance would be misrepresented. This adds a naieve target size and entry overhead to preallocate indexes which are selected at random, to make sure we're going to memory for each iteration.

DanielThomas commented:
@PaulSandoz I realized that at low cardinalities the size of the indexes meant they were just sitting in cache and misrepresenting the actual performance of low cardinality indexes.

Here's how it looks after pre-allocating a bunch of indexes and choosing them at random:

```
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinal               1  avgt       475.050          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinal              10  avgt       332.371          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinal             100  avgt       142.713          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinal            1000  avgt       108.336          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinal           10000  avgt       109.323          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinal          100000  avgt       125.365          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinal         1000000  avgt       190.041          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinalMissing        1  avgt       351.815          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinalMissing       10  avgt       212.487          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinalMissing      100  avgt        99.850          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinalMissing     1000  avgt        78.202          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinalMissing    10000  avgt       134.669          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinalMissing   100000  avgt        31.717          ns/op
HollowPrimaryKeyIndexBenchmark.LoadHollowPrimaryKeyIndexBenchmark.getMatchingOrdinalMissing  1000000  avgt        40.704          ns/op
```

DanielThomas committed:
Fix checkstyle warning

PaulSandoz commented:
Closing as a variant of this was merged with #361
